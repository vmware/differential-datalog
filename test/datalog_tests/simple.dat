echo Metro test;

start;

insert Links(4, "St.Germain", "Odeon"),
insert Links(4, "Odeon", "St.Michel"),
insert Links(4, "St.Michel", "Chatelet"),
insert Links(1, "Chatelet", "Louvres"),
insert Links(1, "Louvres", "Palais Royal"),
insert Links(1, "Palais-Royal", "Tuileries"),
insert Links(1, "Tuileries", "Concorde");

commit;

dump;

echo Rules test;
start;
insert R3(0, true),
insert R3(1, true),
insert R3(2, true),
insert R3(3, true),
insert R3(4, true),
insert R3(5, true),
insert R3(6, true);
commit;
dump;

start;
insert Table1(0, "foo"),
insert Table1(1, "foo"),
insert Table1(0xabcdef000, "foo"),
insert Table1(125, "buzzzzzzzz");
insert Table2("foo", 1234567890, S{(true,true),5}),
insert Table2("foo", 12, S{(true,false),0xabcd}),
insert Table2("foo", 100000000000000, S{(false,true),10000}),
insert Table2("buzzzzzzzz", 1234567890, S{(true,true),5});
commit;
dump Table12;

start;

insert Rel1(0, IP4{100}),
insert Rel1(1, IP6{200}),
insert Rel2(0, Option1{0, IP4{300}, (true, "foo")});

commit;

echo Rel3:;
dump Rel3;

start;

insert AggregateMe1("a", "1"),
insert AggregateMe1("a", "2"),
insert AggregateMe1("a", "3"),
insert AggregateMe1("b", "1"),
insert AggregateMe1("b", "2"),
insert AggregateMe1("b", "3");

commit;

echo Aggregate1;
dump Aggregate1;

echo Aggregate2;
dump Aggregate2;

echo Aggregate3;
dump Aggregate3;

echo Aggregate4;
dump Aggregate4;

echo Disaggregate;
dump Disaggregate;

start;

insert WithKey(0, "foo"),
insert WithKey(1, "bar"),
insert WithKey(2, "buzz");

commit;

echo WithKeyDbg;
dump WithKeyDbg;

start;

delete WithKey(0, "foo");

commit;

echo WithKeyDbg after deletion by value;
dump WithKeyDbg;

start;

delete_key WithKey 1;

commit;

echo WithKeyDbg after deletion by key;
dump WithKeyDbg;


start;

insert WithKey(1, "bar2"),
insert WithKey(3, "xxx"),
insert WithKey(4, "hello"),
delete_key WithKey 2,
delete_key WithKey 3;

commit;

echo WithKeyDbg: +3, -2;
dump WithKeyDbg;

start;

insert Gangster("Billy Jack", "Vito Giacalone"),
insert Gangster("Black Leo", "Leonardo Cellura"),
insert Gangster("Black Sam", "Sam Todaro"),
insert Gangster("Blackie", "James T. Licavoli"),
insert Gangster("Bobby Doyle", "Girolamo Santucci"),
insert Gangster("Bobby Ha Ha", "Robert Attanasio"),
insert Gangster("Bobby Phones", "Robert Sanseverino"),
insert Gangster("Bomp", "Frank Bompensiero"),
insert Gangster("Boo Boo", "Max Hoff"),
insert Gangster("Boobie", "John Cerasini"),
insert Gangster("Bozey", "Arnold Squitieri"),
insert Gangster("The Brain", "Arnold Rothstein"),
insert Gangster("The Bug", "Charles Workman"),
insert Gangster("Bugs", "George Moran"),
insert Gangster("Bugsy", "Benjamin Siegel"),
insert Gangster("The Builder", "Joe Andriacchi"),
insert Gangster("Butsey", "Frank Morelli"),

insert Suspect("John Doe"),
insert Suspect("Bill Smith"),
insert Suspect("Arnold Squitieri"),
insert Suspect("Frank Morelli");

commit;

echo Innocent;
dump Innocent;

start;

insert Address(IPAddr{10,10,10,101}),
insert Address(IPAddr{10,10,10,102}),
insert Address(IPAddr{10,10,10,103}),
insert Address(IPAddr{10,10,10,104}),
insert Address(IPAddr{10,10,10,105}),
insert Address(IPAddr{192,168,0,1}),
insert Address(IPAddr{192,168,0,2}),
insert Address(IPAddr{192,168,0,3}),
insert Address(IPAddr{192,168,0,4}),
insert Address(IPAddr{192,168,0,5}),

insert Blacklist(IPAddr{192,168,0,0}),
insert Blacklist(IPAddr{172,16,0,0});

commit;

echo ValidDestination;
dump ValidDestination;

start;
insert HostAddresses(0, ["10.10.10.101", "10.10.10.102", "10.10.10.103"]),
insert HostAddresses(0, ["10.10.10.103", "10.10.10.104", "10.10.10.105"]),
insert HostAddresses(1, ["192.168.0.1", "192.168.0.2", "192.168.0.3"]);
commit;

dump HostAddress;


start;

insert ExternalIds(0, [("a", "1"),("b", "2"),("c", "3")]),
insert ExternalIds(0, [("c", "4"),("d", "5"),("e", "6")]);
commit;

dump ExternalId;

start;

insert Realized("a", "1", 30),
insert Realized("a", "2", 31),
insert Realized("a", "3", 50),

insert Realized("b", "1", 10000),
insert Realized("b", "2", 100),
insert Realized("b", "3", 16777215),

insert Request("a", "1"),
insert Request("a", "2"),
insert Request("a", "3"),
insert Request("a", "4"),
insert Request("a", "5"),

insert Request("b", "1"),
insert Request("b", "3"),
insert Request("b", "4"),
insert Request("b", "5");



commit;

dump Allocation;
