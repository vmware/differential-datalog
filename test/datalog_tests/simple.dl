/* Test types */

// This is a one-line comment

/* This is a
   multi-line comment */

typedef t00 = bigint
typedef t01 = bool
typedef t02 = string
typedef t03 = bit<32>
typedef t04 = ( bigint )
typedef t05 = ( bigint, bigint )
typedef t06 = ( bigint, bigint, ( bigint ) )
typedef t07 = ()
typedef t08 = ( (), (), () )
typedef t09 = ( (bigint), (), (bigint) )
typedef t10 = ( bigint, bool, string, bit<32> )
typedef t11 = ( bigint, (bool, (string, (bit<32>))))
typedef t12 = t11
typedef t13 = ( t00, t01, t02, t03 )
typedef t14 = ( t04, t04 )
typedef t15 = Const15 { f: bigint }
typedef T16 = t15
typedef T17 = bigint
typedef t18 = Const18 { f0: bigint, f1: string, f2: t00 }
typedef t19 = C000 { f0: bigint }
            | C111 { f0: bigint }
            | C222 { f0: bigint }


typedef person = string

input relation Human(h: person)
input relation Parent(child: person, parent: person)

output relation Sib(s1: person, s2: person)


Sib(x, y) :- Parent(x, z), Parent(y, z), x != y.

Parent("Alice", "Bob").

Parent("Ben", "Bob").

Parent("Bob", "Bob").

/*
Sib("Alice", "Bob").
Sib("Bob", "Alice").
Sib("Ben", "Alice").
Sib("Alice", "Ben").
Sib("Ben", "Bob").
*/

// Tests for datalog functions and expressions

extern function f(): bigint
extern function g(a: bigint): bigint
extern function h(a: (bigint, bigint)): (bigint, bigint)

function a0(): bigint
    = 1

function a1(): bigint
    = 1 + 2

function a2(): bit<32>
    = 32'd0 + 32'd1

function a3(): bit<32>
    = 0: bit<32>

function a4(): bit<32>
    = (0: bit<64>)[40:9]

function a5(a: bit<32>, b: bit<32>): bit<32>
    = (a & b) | (a | b) | (~a) | (a << 32'd5) | (a >> 32'd5) | (a[15:0] ++ a[31:16])

function a6(a: bit<16>, b: bit<16>): bit<16>
    = (a + b) + (a - b) + (a / b) + (a * b) + (a % b)

function a7(a: bit<16>, b: bit<16>): bool
    = (a < b) or (a > b) or (a <= b) or (a >= b) or (a == b) or (a != b) or { a < b }

function a8(): bit<32>
    = 32'd125 | 32'hFF | 32'o777 | 32'b1010101011

function a9(): bit<32>
    = 32'd1_2_5 | 32'hF_F___ | 32'o7_7_7 | 32'b10_1010_1011

function b0(): bool
    = true and false

function b1(a: bool): bool
    = (a and true) or (a or false) or (a => false) or (not a)

function v2(): bool
    = b1(true)

function c0(a: bit<32>, b: bit<16>): bit<16>
    = a; b

function s0(): string
    = "Some string"

function s1(): string
    = "\t\r\n\"\\\a"

function v(): string =
    var v1: string = "hello";
    var v2 = "there";
    v2

typedef Alt = C0{x: bit<32>}
            | C1{x: bit<32>}

function x(): Alt
    = C0{32'd5}

typedef C = C{f1: string, f2: string}

function vars(): () = {
    var x: bigint = 0;
    x = 10;
    var y = C{.f1 = "foo", .f2 = "bar"};
    var z: C = C{.f1 = "bar", .f2 = "foo"};

    (var a, var b) = (x+5, x-5);

    C{.f1 = var e} = y;

    C{var c, var d} = y
}

function patterns(): () = {
    var a: Alt = C0{1};

    var b = match (a) {
        C0{} -> true,
        C1{} -> false
    };

    var i: bit<32> = match (a) {
        C0{.x = v} -> v,
        C1{v} -> v
    }
}

typedef option_t<'A> = None
                     | Some {value : 'A}


function shadow(): string = {
    var b: option_t<string> = None;
    var a = Some{"foo"};
    match (a) {
        Some{v} -> v,
        None    -> ""
    }
}

typedef nested_t = N{field: C}

function fnested(x: nested_t): string = {
    N{C{var res, _}} = x;
    res
}

extern function parameterized(x: 'A, y: 'A): 'A

function parameterized2(x: 'A, y: 'A): bool = x == y

typedef string_syn = string

function use_parameterized(x: string, y: string_syn): string =
    parameterized(x, y)

// metro: graph reachability

typedef line = bigint
typedef station = string

input relation Links(l: line, s1: station, s2: station)

output relation Reach(s1: station, s2: station)


Reach(x, y) :- Links(l, x, y).
Reach(x, y) :- Links(l, y, x).
Reach(x, y):- Links(l, x, z), Reach(z, y).

output relation Answer(x: station)

Answer(x) :- Reach("Odeon", x).

/* Tests for match pattern analysis logic.  Checks that:
   patterns are exhaustive (cover all possible values of the given
   type) and non-redundant. */

typedef T1 = C11{f1: bigint, f2: (bigint, string)}
           | C12{f3: bit<32>, f4: (bool, bool, bool)}

typedef T2 = C21{f1: T1, f2: bigint}
           | C22{f3: (bool, T1)}


typedef T3 = C31{f1: T1, f2: bigint, s: string, b: bit<32>}
           | C32{f3: ((bool, (bool, bool, bool)), T1)}

function f1(x: T1, y: T2, q: T3) : bool = {
    match (x) {
        C11 -> true,
        _   -> false
    };

    match (x) {
        C11 -> true,
        C12 -> false
    };

    match (y) {
        C21{_,_}        -> true,
        C22{(_, C11{})} -> true,
        _               -> false
    };

    match (y) {
        C21{_,_}        -> true,
        C22{(true, C11{})} -> true,
        C22{(true, C12{})} -> true,
        C22{(false, C11{})} -> true,
        C22{(false, C12{})} -> true
    };


    var z: (bool, (bool, (bool,bool))) = (true, (false, (true, false)));

    match (z) {
        (true, _)                   -> true,
        (false, (_,(true, true)))   -> true,
        (false, (_,(true, false)))  -> true,
        (false, (_,(false, true)))  -> true,
        (false, (_,(false, false))) -> true
    };

    match (z) {
        (_, (_,(true, true)))       -> true,
        (_, (_,(true, false)))      -> true,
        (_, (_,(false, true)))      -> true,
        (false, (_,(false, false))) -> true,
        (true, (_,(false, false)))  -> true
    };

    match (q) {
        C31{}                                                      -> true,
        C32{.f3=((true, (_, _, false)), _)}                        -> true,
        C32{.f3=((false, (_, _, false)), _)}                       -> true,
        C32{.f3=((false, (_, _, true)), _)}                        -> true,
        C32{.f3=((_, (_, _, true)), C11{})}                        -> true,
        C32{.f3=((_, (_, _, true)), C12{.f4=(true, true, true)})}  -> false,
        C32{.f3=((_, (_, _, true)), C12{.f4=(_, _, false)})}       -> false,
        C32{.f3=((_, (_, _, true)), C12{.f4=(_, _, true)})}        -> false
    };

    match (q) {
        C31{_, 0, "foo", 15}                                       -> true,
        C31{_, _, _, 15}                                           -> true,
        C31{_, 0, _, _}                                            -> true,
        C31{_, _, "foo", _}                                        -> true,
        C32{}                                                      -> true,
        C31{_, _, _, _}                                            -> true
    };

    var s: string = "bar";
    match (s) {
        "foo" -> true,
        "bar" -> true,
        _     -> false
    }
}


/* Statements test */
input relation X(field: bigint)
output relation Z(field: bigint)

output relation W(f1: bigint, f2: bigint, f3: bigint)

for (a in X)
   skip

for (a in X)
   Z(a.field)
// equivalent to
Z(a.field) :- X[a].

for (a in X)
   match (a.field) {
      1 -> Z(a.field),
      2 -> Z(a.field+1),
      _ -> Z(a.field-1)
   }

for (X{a} in X)
   match (a) {
      1 -> Z(a),
      2 -> Z(a+1),
      _ -> Z(a-1)
   }
// both are equivalent to
Z(a) :- X(a), match(a) { 1 -> true, _ -> false }.
Z(a+1) :- X(a), match(a) { 1-> false, 2 -> true, _ -> false }.
Z(a-1) :- X(a), match(a) { 1 -> false, 2 -> false, _ -> true }.

for (a in X if true)
   skip
// equivalent to
for (a in X)
   if (true)
       skip

for (a in X) {
   skip;
   skip
}

for (a in X if (a.field % 2) == 0)
    Z(a.field)
// equivalent to
for (a in X)
    if (a.field % 2 == 0)
         Z(a.field)
// equivalent to
Z(a) :- X(a), ((a % 2) == 0).

for (a in X) {
    if (a.field == 2)
        skip
}

for (a in X) {
    if (a.field == 2)
        skip
    else
        skip
}

for (a in X) {
    if (a.field == 2)
        Z(a.field+1)
    else
        Z(a.field)
}
// equivalent to
Z(a+1) :- X(a), a == 2.
Z(a) :- X(a), a != 1.

for (a in X)
    if (a.field >= 2)
        if (a.field > 2) {
            Z(a.field);
            Z(a.field+2)
        }
// equivalent to
Z(a) :- X(a), a>=2, a > 2.
Z(a+2) :- X(a), a>=2, a > 2.

for (a in X) {
    for (b in X)
        Z(a.field+b.field);
    for (c in X)
        Z(a.field+c.field)
}
// equivalent to
Z(a+b) :- X(a), X(b).
Z(a+c) :- X(a), X(c).
// Equivalent to
Z(a+b), Z(a+c) :- X(a), X(b), X(c).

for (a in X)
   var b = a.field+2 in
       skip

for (a in X)
   var b: bigint = a.field+2 in
       Z(a.field+b)
// equivalent to
Z(a+b) :- X(a), var b = a+2.

for (a in X)
   var b = a.field+2, c = b+a.field in
       W(a.field, b, c)
// equivalent to
W(a, b, c) :- X(a), var b = a+2, var c = b+a.

/* simple rules */

typedef R1 = bigint

input relation R1[R1]
input relation R2[(bigint, bigint)]

R1[1].
R2[(1,2)].

input relation R3(f1: bigint, f2: bool)
output relation R4[R3]
output relation R5[bigint]
output relation R6(f: bigint)

R4[x] :- R3[x].
R5[x.f1] :- R3[x].
R6(x.f1) :- R3[x].

input relation XX(b: bigint, c: R3, d: R1)
output relation ZZ(x: bigint, y: R3)
ZZ(a.b, a.c) :- XX[a].

input relation Y(b: bigint, e: string)
output relation T(x: bigint, y: string)

T(a.b, e) :- XX[a], Y(a.d, e).

typedef R7 = R7{f1: bigint, f2: bit<16>}
           | R7Other

output relation R7[R7]

R7(1,2).
R7[R7{1,3}].
R7[R7Other].

typedef S = S {
    f1: (bool, bool),
    f2: bit<32>
}
input relation Table1(id: bit<64>, name: string)
input relation Table2(name: string, f2: bigint, f3: S)

output relation Table12(id: bit<128>, name: string, f3: S)

Table12(hash128((id,name,f2,f3)), name, f3) :- Table1(id, name), Table2(name, f2, f3).


/* strings test */
input relation Strings(s: string)

Strings("").
Strings("foo").
Strings("bar\n").
Strings("\t\tbar").
Strings("\"buzz").
Strings("\r\n\abuzz").
Strings("\\buzz").
Strings("\\buzz").

/* unicode */
Strings("barΔ").

/* raw strings */
Strings([||]).
Strings([|foo|]).
Strings([|foo
buzz
bar|]).

Strings([|
    foo
    buzz
    bar|]).

Strings([|barΔ
Π|]).

Strings([|barΔ
Π|] " quoted string").

// backslash does not work as an escape character in raw strings
Strings([|
    foo\r\n
    buzz
    bar|]).

/* string literal concatenation */
Strings("foo" "-" "bar").

Strings("foo\n" "bar").
Strings("foo\\""bar").

Strings("foo\\" [|bar|]).

function dummy(x: string): string = x ++ " dummy"

function strings(): () = {
    var str1 = "foo" ++ "bar";
    var str2 = "foo" "bar" ++ "buzz" ++ [|
raw string |]"quoted string";
    var str3 = $"str1: ${ str1 }";
    var str4 = $"str1: ${str1}, str2: ${str2}, str3: ${str3}.";
    var str5 = $[|str1: ${str1},
str2: ${str2},
str3: ${str3}.|];
    var str6 = $"${str1 ++ str2}";
    var str7 = $"${{var x = str1 ++ str2; x}}";
    var str8 = $"${{var x = str1 ++ str2; dummy(x)}}";
    ()
}

Strings("word1\
        \ word2").

Strings("line1\n\
        \line2").

function tostring1(): () = {
    var a: bigint  = 5;
    var b: bit<32> = 5;
    var c: bool    = true;
    var d: string  = "foo";
    var err = $"a=${a}, b=${b}, c=${c}, d=${d}";
    ()
}

typedef serializable_t = ConsInt {x: bigint}
                       | ConsBit {y: bit<32>}
                       | ConsBool{z: bool}
                       | Cons0{}

function serializable_t2string(x: serializable_t): string = {
    match (x) {
        ConsInt{v}      -> "ConsInt(" ++ v ++ ")",
        ConsBit{v}      -> $"ConsBool(${v})",
        ConsBool{true}  -> $"ConsBool(true)",
        ConsBool{false} -> $"ConsBool(false)",
        Cons0{}         -> $"Cons0"
    }
}

function tostring2(): () = {
    var a: bigint     = 5;
    var b: bit<32> = 5;
    var c: bool    = true;
    var d: string  = "foo";
    var e: serializable_t = ConsInt{0};
    var err = $"a=${a}, b=${b}, c=${c}, d=${d}, e:${e}";
    ()
}

/* More random examples */
typedef IP4Addr = bit<32>
typedef IP6Addr = bit<128>

typedef IP4Subnet = IP4Subnet {
    addr: IP4Addr,
    mask: bit<32>
}

typedef IPAddr = IP4{ip4: IP4Addr}
               | IP6{ip6: IP6Addr}

typedef FooStruct = Option1 {f1: bigint, f2: IPAddr, f3: (bool, string)}
                  | Option2 {f4: bit<32>, f5: IPAddr}

function foo(): bool = {
    var v1: bigint             = 0;
    var v2: bit<32>         = 1;
    var v3: bit<129>        = 129'd0;
    var v4: bool            = true;
    var v5: (bigint, string)   = (0, "test_string");
    var v6 = IP4{32'haabbccdd};
    var v7 = Option1{.f1 = 0, .f2 = IP4{32'd0}, .f3 = (true, "bar")};

    v3 = v3 + 129'habcde;

    match (v6) {
        IP4{_} -> true,
        _      -> false
    };

    match (v7) {
        Option1{.f2 = IP6{_}, .f3 = (_, "bar")} -> true,
        _                                       -> false
    }
}

input relation Rel1(x: bigint, y: IPAddr)
input relation Rel2(x: bigint, z: FooStruct)

output relation Rel3(x:bigint, y: IPAddr, z: FooStruct)

Rel3(x, y, z) :- Rel1(x, y), Rel2(x,z).


output relation Rel4(x:bigint, y: IPAddr, b: bool)

Rel4(x,y,b) :- Rel1(x,y),
               Rel2(x, Option1{.f2=IP4{}, .f3=(b, "buzz")}).

/* Aggregation */

input relation AggregateMe1(x: string, y: string)

output relation Aggregate1(x: string, cnt: bit<64>)

Aggregate1(x, cnt) :- AggregateMe1(x,y), Aggregate((x), cnt = count(y)).

output relation Aggregate2(x: string, set: Set<string>)
Aggregate2(x, set) :- AggregateMe1(x,y), Aggregate((x), set = group2set(y)).

output relation Aggregate3(x: string, vec: Vec<string>)
Aggregate3(x, vec) :- AggregateMe1(x,y), Aggregate((x), vec = group2vec(y)).


output relation Aggregate4(x: string, map: Map<string,string>)
Aggregate4(x, map) :- AggregateMe1(x,y), Aggregate((x), map = group2map((x,y))).

output relation Disaggregate(x: string, y: string)
Disaggregate(x,y) :- AggregateMe1(x,y), Aggregate((x), set = group2set(y)), var y = FlatMap(set).

/* Primary key */
input relation WithKey(key: bit<128>, val: string)
primary key (x) x.key

output relation WithKeyDbg(key: bit<128>, val: string)
WithKeyDbg(k,v) :- WithKey(k, v).

/* More advanced antijoins */

input relation Gangster(nickname: string, name: string)
input relation Suspect(name: string)
output relation Innocent(name: string)

Innocent(name) :- Suspect(name), not Gangster(_, name).

typedef ip_addr_t = IPAddr{b3: bit<8>, b2: bit<8>, b1: bit<8>, b0: bit<8>}

input relation Blacklist(addr: ip_addr_t)
input relation Address(addr: ip_addr_t)

output relation ValidDestination(addr: ip_addr_t)
ValidDestination(addr) :-
    Address(addr),
    IPAddr{var b3, var b2, _, _} = addr,
    not Blacklist(IPAddr{b3, b2, _, _}).

// FlatMap test
input relation HostAddresses(host: bit<64>, addrs: Set<string>)
output relation HostAddress(host: bit<64>, addr: string)

HostAddress(host, addr) :- HostAddresses(host, addrs),
                           var addr = FlatMap(addrs).


input relation ExternalIds(host: bit<64>, ids: Map<string, string>)
output relation ExternalId(host: bit<64>, id: (string, string))

ExternalId(host, id) :- ExternalIds(host, ids),
                        var id = FlatMap(ids).

/* Allocator test */

extern function allocate_u32(allocated: Set<bit<32>>, toallocate: Vec<'B>, max_val: bit<32>): Vec<('B, bit<32>)>

// allocate a unique x per name
input relation Realized(name: string, id: string, x: bit<32>)
input relation Request(name: string, id: string)

output relation Allocation(name: string, id: string, x: bit<32>)

Allocation(name, id, x) :-
    Request(name, id), Realized(name, id, x).

relation ToAllocate(name: string, ids: Vec<string>)
ToAllocate(name, ids) :-
    Request(name, id), not Realized(name, id, _), Aggregate((name), ids = group2vec(id)).

relation Allocated(name: string, xs: Set<bit<32>>)
Allocated(name, xs) :-
    Realized(name, _, x), Aggregate((name), xs = group2set(x)).

Allocation(name, id, x) :-
    ToAllocate(name, ids),
    Allocated(name, xs),
    var allocation = FlatMap(allocate_u32(xs, ids, (32'd1 << 32'd24) - 32'd1)),
    (var id, var x) = allocation.
