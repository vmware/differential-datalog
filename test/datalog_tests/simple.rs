fn parameterized<A: Val>(x: &A, y: &A) -> A {
    x.clone()
}

fn allocate_u32<B: Ord+Clone>(allocated: &std_Set<u32>, toallocate: &std_Vec<B>, max_val: &u32) -> std_Vec<(B,u32)> {
    let first = *allocated.x.range(..).next_back().unwrap_or(&0);
    let mut last = first;
    let mut res = std_Vec::new();
    for b in toallocate.x.iter() {
        loop {
            last = if last + 1 <= *max_val {last + 1} else {1};
            if last == first {
                return res;
            };
            if allocated.x.contains(&last) { continue; };
            res.x.push(((*b).clone(), last));
            break;
        }
    };
    return res
}
