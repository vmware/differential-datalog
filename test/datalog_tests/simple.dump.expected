Metro test
Answer:
Answer{"Chatelet"}
Answer{"Louvres"}
Answer{"Odeon"}
Answer{"Palais Royal"}
Answer{"St.Germain"}
Answer{"St.Michel"}

Reach:
Reach{"Chatelet","Chatelet"}
Reach{"Chatelet","Louvres"}
Reach{"Chatelet","Palais Royal"}
Reach{"Chatelet","St.Michel"}
Reach{"Concorde","Tuileries"}
Reach{"Louvres","Chatelet"}
Reach{"Louvres","Louvres"}
Reach{"Louvres","Palais Royal"}
Reach{"Odeon","Chatelet"}
Reach{"Odeon","Louvres"}
Reach{"Odeon","Odeon"}
Reach{"Odeon","Palais Royal"}
Reach{"Odeon","St.Germain"}
Reach{"Odeon","St.Michel"}
Reach{"Palais Royal","Louvres"}
Reach{"Palais-Royal","Concorde"}
Reach{"Palais-Royal","Palais-Royal"}
Reach{"Palais-Royal","Tuileries"}
Reach{"St.Germain","Chatelet"}
Reach{"St.Germain","Louvres"}
Reach{"St.Germain","Odeon"}
Reach{"St.Germain","Palais Royal"}
Reach{"St.Germain","St.Germain"}
Reach{"St.Germain","St.Michel"}
Reach{"St.Michel","Chatelet"}
Reach{"St.Michel","Louvres"}
Reach{"St.Michel","Odeon"}
Reach{"St.Michel","Palais Royal"}
Reach{"St.Michel","St.Michel"}
Reach{"Tuileries","Concorde"}
Reach{"Tuileries","Palais-Royal"}
Reach{"Tuileries","Tuileries"}

Rules test
Answer:
Answer{"Chatelet"}
Answer{"Louvres"}
Answer{"Odeon"}
Answer{"Palais Royal"}
Answer{"St.Germain"}
Answer{"St.Michel"}

R4:
R3{0,true}
R3{1,true}
R3{2,true}
R3{3,true}
R3{4,true}
R3{5,true}
R3{6,true}

R5:
0
1
2
3
4
5
6

R6:
R6{0}
R6{1}
R6{2}
R6{3}
R6{4}
R6{5}
R6{6}

Reach:
Reach{"Chatelet","Chatelet"}
Reach{"Chatelet","Louvres"}
Reach{"Chatelet","Palais Royal"}
Reach{"Chatelet","St.Michel"}
Reach{"Concorde","Tuileries"}
Reach{"Louvres","Chatelet"}
Reach{"Louvres","Louvres"}
Reach{"Louvres","Palais Royal"}
Reach{"Odeon","Chatelet"}
Reach{"Odeon","Louvres"}
Reach{"Odeon","Odeon"}
Reach{"Odeon","Palais Royal"}
Reach{"Odeon","St.Germain"}
Reach{"Odeon","St.Michel"}
Reach{"Palais Royal","Louvres"}
Reach{"Palais-Royal","Concorde"}
Reach{"Palais-Royal","Palais-Royal"}
Reach{"Palais-Royal","Tuileries"}
Reach{"St.Germain","Chatelet"}
Reach{"St.Germain","Louvres"}
Reach{"St.Germain","Odeon"}
Reach{"St.Germain","Palais Royal"}
Reach{"St.Germain","St.Germain"}
Reach{"St.Germain","St.Michel"}
Reach{"St.Michel","Chatelet"}
Reach{"St.Michel","Louvres"}
Reach{"St.Michel","Odeon"}
Reach{"St.Michel","Palais Royal"}
Reach{"St.Michel","St.Michel"}
Reach{"Tuileries","Concorde"}
Reach{"Tuileries","Palais-Royal"}
Reach{"Tuileries","Tuileries"}

Table12{20695836920908937261915266253713563738,"foo",S{(true, true),5}}
Table12{81692126345696542188923640651411714412,"foo",S{(false, true),10000}}
Table12{137151609739491206170748566711757625861,"foo",S{(true, false),43981}}
Table12{173592105249357153805573863620167017533,"foo",S{(false, true),10000}}
Table12{173670172741911619502105830072953828992,"foo",S{(true, false),43981}}
Table12{211727904196777888792921590151204575764,"foo",S{(true, true),5}}
Table12{218334107078988268330177571792522447425,"foo",S{(true, true),5}}
Table12{258101594679877739861006822360175003366,"buzzzzzzzz",S{(true, true),5}}
Table12{312560568469280405643575197132576203606,"foo",S{(false, true),10000}}
Table12{324905179504175058909924909437678210347,"foo",S{(true, false),43981}}
Rel3:
Rel3{0,IP4{100},Option1{0,IP4{300},(true, "foo")}}
Aggregate1
Aggregate1{"a",3}
Aggregate1{"b",3}
Aggregate2
Aggregate2{"a",std_Set { x: {"1", "2", "3"} }}
Aggregate2{"b",std_Set { x: {"1", "2", "3"} }}
Aggregate3
Aggregate3{"a",std_Vec { x: ["1", "2", "3"] }}
Aggregate3{"b",std_Vec { x: ["1", "2", "3"] }}
Aggregate4
Aggregate4{"a",std_Map { x: {"a": "3"} }}
Aggregate4{"b",std_Map { x: {"b": "3"} }}
Disaggregate
Disaggregate{"a","1"}
Disaggregate{"a","2"}
Disaggregate{"a","3"}
Disaggregate{"b","1"}
Disaggregate{"b","2"}
Disaggregate{"b","3"}
WithKeyDbg
WithKeyDbg{0,"foo"}
WithKeyDbg{1,"bar"}
WithKeyDbg{2,"buzz"}
WithKeyDbg after deletion by value
WithKeyDbg{1,"bar"}
WithKeyDbg{2,"buzz"}
WithKeyDbg after deletion by key
WithKeyDbg{2,"buzz"}
WithKeyDbg: +3, -2
WithKeyDbg{1,"bar2"}
WithKeyDbg{4,"hello"}
Innocent
Innocent{"Bill Smith"}
Innocent{"John Doe"}
ValidDestination
ValidDestination{IPAddr{10,10,10,101}}
ValidDestination{IPAddr{10,10,10,102}}
ValidDestination{IPAddr{10,10,10,103}}
ValidDestination{IPAddr{10,10,10,104}}
ValidDestination{IPAddr{10,10,10,105}}
HostAddress{0,"10.10.10.101"}
HostAddress{0,"10.10.10.102"}
HostAddress{0,"10.10.10.103"}
HostAddress{0,"10.10.10.104"}
HostAddress{0,"10.10.10.105"}
HostAddress{1,"192.168.0.1"}
HostAddress{1,"192.168.0.2"}
HostAddress{1,"192.168.0.3"}
ExternalId{0,("a", "1")}
ExternalId{0,("b", "2")}
ExternalId{0,("c", "3")}
ExternalId{0,("c", "4")}
ExternalId{0,("d", "5")}
ExternalId{0,("e", "6")}
Allocation{"a","1",30}
Allocation{"a","2",31}
Allocation{"a","3",50}
Allocation{"a","4",51}
Allocation{"a","5",52}
Allocation{"b","1",10000}
Allocation{"b","3",16777215}
Allocation{"b","4",1}
Allocation{"b","5",2}
